app/controllers/course/survey/responses_controller.rb

```diff
@@ -24,20 +24,36 @@ def create
   end
 
   def show
-    render 'course/survey/surveys/index'
+    respond_to do |format|
+      format.html { render 'course/survey/surveys/index' }
+      format.json { render_response_json }
+    end
   end
 
   def edit
-    @response.build_missing_answers_and_options
-    if @response.save
-      render_response_json
-    else
-      head :internal_server_error
+    raise CanCan::AccessDenied if cannot?(:submit, @response) && cannot?(:modify, @response)
```

Comment:
👍 As discussed offline,  because `allow_modify_after_submit`  and `allow_response_after_end` are independent, it's not possible to merge the permission.